doc/cprover-manual/contracts-assigns.md

```diff
@@ -1,75 +1,368 @@
-and `__CPROVER_havoc_slice` does not support havocing pointers. 
-### Parameters
+### Object slice targets
+
+```c
+void __CPROVER_object_upto(void *ptr, __CPROVER_size_t size);
+```
+
+Given a pointer `ptr` pointing into some object (possibly at some non-zero offset),
+`__CPROVER_object_upto(ptr, size)` specifies that the range of `size` bytes starting
+at `ptr` is assignable.
 
-An _assigns_ clause currently supports simple variable types and their pointers,
-structs, and arrays.  Recursive pointer structures are left to future work, as
-their support would require changes to CBMC's memory model.
+
+The value of `size` must such that the range does not exceed
+the object's boundary, i.e.
+`size <= __CPROVER_OBJECT_SIZE(ptr) - __CPROVER_POINTER_OFFSET(ptr)` must hold.
+
+In the example below, the `struct vect_t`, its `data` array and an exta hidden
+byte are packed together in a single object by the `vec_alloc` function.
+The `vec_clear` function can only assign `vec->size` bytes starting from
+`vec->data`. As a result the assignments to `vec->size` and the hidden byte
+fail the verification.
 
 ```c
-/* Examples */
-int err_signal; // Global variable
+#include <stdlib.h>
 
-int sum(const uint32_t a, const uint32_t b, uint32_t* out)
-__CPROVER_assigns(*out)
+#define MAX_SIZE 10
 
-int sum(const uint32_t a, const uint32_t b, uint32_t* out)
-__CPROVER_assigns(err_signal)
+struct vec_t {
+  size_t size;
+  char *data;
+};
 
-int sum(const uint32_t a, const uint32_t b, uint32_t* out)
-__CPROVER_assigns(*out, err_signal)
+// Allocates a vect_t struct together with its data and a hidden byte
+// in a same object.
+struct vec_t *vec_alloc(size_t size) {
+
+  if(size > MAX_SIZE)
+    size = MAX_SIZE;
+
+  // allocate the struct + data + 1 extra hidden byte
+  struct vec_t *vec = malloc(sizeof(struct vec_t) + size + 1);
+  if (vec) {
+    vec->size = size;
+    vec->data = ((char *)vec) + sizeof(struct vec_t);
+  }
+  return vec;
+}
+
+// Clear the vec->data array
+void vec_clear(struct vec_t *vec)
+  __CPROVER_assigns(
+    vec && vec->data: __CPROVER_object_upto(vec->data, vec->size))
+{
+  if (!vec)
+    return;
+
+  vec->size = vec->size; // FAILURE
+
+  for (size_t i = 0; i < vec->size; i++)
+    vec->data[i] = 0; // SUCCESS
+
+  char *hidden_byte = ((char *)vec + sizeof(*vec) + vec->size);
+  *hidden_byte = 0; // FAILURE
+}
+
+// Proof harness
+int main() {
+  size_t size;
+  struct vec_t *vec = vec_alloc(size);
+  vec_clear(vec);
+}
+```
+
+---
+
+```c
+void __CPROVER_object_from(void *ptr);
+```
+
+Given a pointer `ptr` pointing into some object (possibly at some non-zero offset),
+`__CPROVER_object_from(ptr)` specifies that the range of bytes starting at `ptr`
+and of size `__CPROVER_OBJECT_SIZE(ptr) - __CPROVER_POINTER_OFFSET(ptr)` is assignable.
+
+Revisiting our previous example, changing the target to
+`__CPROVER_object_from(vec->data)` still rejects the assignment to vec->size,
```

Comment:
"vec->size" should be formatted as code.

Reply:
fixed